File: problem_16.cpp
#include <chrono>
#include <iostream>
#include <cmath>
#include <cstring>
#include <vector>
#include <array>

const int maxExp = 305;
const int base = 2;
const int numFactors = 6;
const int maxDoublings = 10;
const int digitMaxVal = 10;
std::array<int, numFactors> usedSteps = { 3, 5, 6, 7, 8, 9 };

std::vector<unsigned char> multiply_strings (std::vector<unsigned char> f1, std::vector<unsigned char> f2) 
{
    std::vector<unsigned char> res;
    res.resize(f1.size() + f2.size() + 1);
    memset(res.data(), 0, res.size() * sizeof(unsigned char));
    for (int i = 0; i < f1.size(); i++) 
    {
        for (int j = 0; j < f2.size(); j++) 
        {
            int resPos = i + j;
            int digitMult = f1.at(i) * f2.at(j);
            res.at(resPos) += (digitMult) % digitMaxVal;
            if (digitMult >= digitMaxVal) 
            {
                res.at(resPos + 1) += int(digitMult / digitMaxVal);
            }
            for (int k = resPos; k < res.size(); k++) 
            {
                int rip = int(res.at(k) / digitMaxVal);
                if (rip >= 1) 
                {
                    res.at(k) -= rip * digitMaxVal;
                    res.at(k + 1) += rip;
                }
                else 
                {
                    break;
                }
            }
        }
    }
    while (res.size() > 0 && res.at(res.size() - 1) == 0) 
    {
        res.resize(res.size() - 1);
    }
    return res;
}

std::vector<unsigned char> compute_result () 
{
    std::vector<std::vector<unsigned char>> res;
    std::vector<unsigned char> workingNum = { (unsigned char) base };
    for (int i = 1; i < maxDoublings; i++) 
    {
        workingNum = multiply_strings(workingNum, workingNum);
        for (int j = 0; j < numFactors; j++)
        {
            if (i == usedSteps.at(j))
            {
                res.push_back(workingNum);
                break;
            }
        }
    }
    std::vector<unsigned char> acc = { (unsigned char) 1 };
    for (int i = 0; i < (int) res.size(); i++)
    {
        acc = multiply_strings(acc, res.at(i));
    }
    return acc;
}

int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();

    int runningSum = 0;
    std::vector<unsigned char> res = compute_result();
    for (int i = 0; i < res.size(); i++) 
    {
        runningSum += (int) res.at(i);
        std::cout << int(res.at(res.size() - 1 - i));
    }
    std::cout << std::endl;
    
    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << runningSum << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_16.rs
use std::time::{Instant};

const BASE: u8 = 2;
const NUM_FACTORS: usize = 6;
const MAX_DOUBLINGS: usize = 10;
const DIGIT_MAX_VAL: u8 = 10u8;
const USED_STEPS: [usize; NUM_FACTORS] = [ 3, 5, 6, 7, 8, 9 ];

fn multiply_strings (f1 : &Vec<u8>, f2: &Vec<u8>) -> Vec<u8>
{
    let mut res: Vec<u8> = vec![];
    res.resize(f1.len() + f2.len() + 1, 0);
    for i in 0..f1.len()
    {
        for j in 0..f2.len()
        {
            let res_pos: usize = i + j;
            let digit_mult: u8 = f1[i] * f2[j];
            res[res_pos] = res[res_pos] + ((digit_mult) % DIGIT_MAX_VAL);
            if digit_mult >= DIGIT_MAX_VAL 
            {
                res[res_pos + 1] = res[res_pos + 1] + (digit_mult / DIGIT_MAX_VAL);
            }
            for k in res_pos..res.len()
            {
                let rip: u8 = res[k] / DIGIT_MAX_VAL;
                if rip >= 1
                {
                    res[k] = res[k] - rip * DIGIT_MAX_VAL;
                    res[k + 1] = res[k + 1] + rip;
                }
                else 
                {
                    break;
                }
            }
        }
    }
    while res.len() > 0 && res[res.len() - 1] == 0
    {
        res.resize(res.len() - 1, 0);
    }
    return res;
}

fn compute_result () -> Vec<u8>
{
    let mut res: Vec<Vec<u8>> = vec![];
    let mut working_num: Vec<u8> = vec![ BASE ];
    for i in 1..MAX_DOUBLINGS
    {
        working_num = multiply_strings(&working_num, &working_num);
        for j in 0..NUM_FACTORS
        {
            if i == USED_STEPS[j]
            {
                res.push(working_num.clone());
                break;
            }
        }
    }
    return res.iter().fold(vec![ 1u8 ], |acc, x| multiply_strings(&acc, x))
}

fn main () 
{
    let start = Instant::now();

    let res: Vec<u8> = compute_result();
    let running_sum: usize = res.iter().fold(0, |acc, x| acc + (*x as usize));
    
    println!("Result: {}", running_sum);
    let duration = start.elapsed();
    println!("Timing: {}", duration.as_nanos());
}




